pyterrier_services/google.py

```diff
@@ -0,0 +1,138 @@
+            raise ValueError("A Google CX must be specified")
+        try:
+            from googleapiclient.discovery import build
+        except ModuleNotFoundError as mnfe:
+            raise Exception("You need to pip install google-api-python-client") from mnfe
+
+        service = build(
+            "customsearch", "v1", developerKey=self.key
+        )
+        def _search_google(one_query : pt.model.IterDict) -> pt.model.IterDict:
+            res = service.cse().list(q=next(one_query)["query"], cx=cx).execute()
+            return res["items"]
+
+        return pt.apply.by_query(_search_google, iter=True)
+
+
+class GoogleSearchRetriever(pt.Transformer):
+    """A :class:`~pyterrier.Transformer` retriever that queries the Google search API."""
+
+    def __init__(self,
+        api: Optional[GoogleApi] = None,
+        cx: Optional[str] = None,
+        *,
+        num_results: int = 10,
+        verbose: bool = False,
+    ):
+        """
+        Args:
+            api (GoogleApi): The Google API service. Defaults to a new instance of :class:`GoogleApi`.
+            cx: (str): The Google Custom Search Engine ID. This is required to perform searches.
+            num_results (int): The number of results to retrieve per query. Defaults to 10.
+            verbose (bool): Whether to log the progress. Defaults to False.
+        """
+        self.api = api or GoogleApi()
+        if cx is None:
+            cx = os.getenv("GOOGLE_CSE_CX")
+        if cx is None:
+            raise ValueError(f"A Google Custom Search Engine ID (cx) must be specified. See <{_HELP_URL}> for details on how to get a Custom Search Engine ID.")
+        self.cx = cx
+        self.cse_service = self.api._build("customsearch", "v1", developerKey=self.api.api_key).cse()
+        self.num_results = num_results
+        self.verbose = verbose
+
+    def transform(self, inp: pd.DataFrame) -> pd.DataFrame:
+        return multi_query(
+            paginated_search(self._search_internal, num_results=self.num_results),
+            verbose=self.verbose,
+            verbose_desc='GoogleSearchRetriever',
+        )(inp)
+
+    def _search_internal(self,
+        query: str,
+        *,
+        offset: int = 0,
+        limit: int = 10,
+        return_next: bool = False,
+        return_total: bool = False
+    ) -> Union[pd.DataFrame, Tuple[pd.DataFrame, int], Tuple[pd.DataFrame, int, int]]:
+        """Searches for papers on Google with the provided query.
+
+        Args:
+            query: The search query.
+            offset: The offset of the first result to retrieve. Defaults to 0.
+            limit: The maximum number of results to retrieve. Defaults to 10.
+            return_next: Whether to return the next query URL. Defaults to False.
+            return_total: Whether to return the total number of results. Defaults to False.
+        """
+        api_result = self.cse_service.list(q=query, cx=self.cx, num=min(limit, 10), start=offset).execute()
```

Comment:
is 10 the maximum number of results?

Reply:
I see multiple requests are used for results more than 10.

Reply:
yep, and an error is raised when num>10